File: java/3447. Assign Elements to Groups with Constraints.java
class Solution {
    public int[] assignElements(int[] groups, int[] elements) {
        int[] answer = new int[groups.length];
        Map<Integer, Integer> elementIndex = new HashMap<>();
        Arrays.fill(answer, Integer.MAX_VALUE);
        for (int i = 0; i < elements.length; i++) {
            elementIndex.putIfAbsent(elements[i], i);
        }
        for (int i = 0; i < groups.length; i++) {
            for (int div : divisors(groups[i])) {
                if (!elementIndex.containsKey(div)) continue;
                answer[i] = Math.min(answer[i], elementIndex.get(div));
            }
            if (answer[i] == Integer.MAX_VALUE) answer[i] = -1;
        }
        return answer;
    }
    private List<Integer> divisors(int num) {
        List<Integer> answer = new ArrayList<>();
        for (int i = 1; i*i <= num; i++) {
            if (num % i == 0) {
                answer.add(i);
                answer.add(num/i);
            }
        }
        return answer;
    }
}

File: python/3447. Assign Elements to Groups with Constraints.py
class Solution:
    def assignElements(self, groups: List[int], elements: List[int]) -> List[int]:

        def divisors(num):
            for i in range(1, num+1):
                if i*i > num:
                    break
                if num % i == 0:
                    yield i
                    yield num/i

        answer = [float('inf')] * len(groups)
        element_index = {}
        for i, ele in enumerate(elements):
            element_index.setdefault(ele, i)
        for i in range(len(groups)):
            for div in divisors(groups[i]):
                if div not in element_index:
                    continue
                answer[i] = min(answer[i], element_index[div])
            if answer[i] == float('inf'):
                answer[i] = -1
        return answer
